examples/custom_search_method/core_search_runner/asha.py

```diff
@@ -0,0 +1,483 @@
+"""
+This example shows ASHA (Asynchronous Successive Halving Algorithm) implemented as a custom SearchMethod.
+For details related to ASHA see https://docs.determined.ai/latest/training/hyperparameter/search-methods/hp-adaptive-asha.html
+
+ASHASearchMethod provides implementation for abstract methods from SearchMethod class, which
+are invoked by CoreSearchRunner in response to the SearcherEvents received
+from the multi-trial experiment. The methods return a list of Operations to CoreSearchRunner
+which sends them to the multi-trial experiment for execution, as depicted below.
+
+Multi-trial experiment  --- (SearcherEvent1) ----> CoreSearchRunner  --- (SearcherEvent1) ---> SearchMethod
+Multi-trial experiment <---   (Operations1)  ----  CoreSearchRunner <---   (Operations1)  ---  SearchMethod
+Multi-trial experiment  --- (SearcherEvent2) ----> CoreSearchRunner  --- (SearcherEvent2) ---> SearchMethod
+Multi-trial experiment <---   (Operations2)  ----  CoreSearchRunner <---   (Operations2)  ---  SearchMethod
+and so on.
+
+Currently, we support the following operations:
+  -> Create - starts a new trial with a unique trial id and a set of hyperparameters,
+  -> ValidateAfter - sets number of steps (i.e., batches or epochs) after which a validation is run for a trial
+                     with a given id,
+  -> Close - closes a trial with a given id,
+  -> Shutdown - closes the experiment.
+
+To support fault tolerance, a custom SearchMethod has to implement save_method_state() and load_method_state(), which
+provide logic for saving and loading any information internal to the SearchMethod,
+such as variables, structures, or models, that are modified as the SearchMethod progresses.
+save_method_state() and load_method_state() are called by CoreSearchRunner as a part of CoreSearchRunner.save() and
+CoreSearchRunner.load() to ensure that CoreSearchRunner state and SearchMethod state can be restored if
+CoreSearchRunner is terminated or interrupted.
+To provide fault tolerance, CoreSearchRunner calls save() and save_method_state() after processing
```

Comment:
ln 29 to 34 seem a little too detailed for this tutorial example.  i suggest removing

Reply:
agree. however, I think similar explanation should happen in docs, to make sure users understand what is happening and won't get creative with save/load methods.